### DISCUSSION TITLE
[@types/jest] Improve matching types

### DISCUSSION THREAD

**Discussion by tomhicks** [DefinitelyTyped/DefinitelyTyped#62128] [Issues with a @types package]
I'm sure there's a reason this hasn't been done, but I always patch this locally in my projects:

```
toMatchObject<S>(expected: T extends Array<infer S> ? Array<DeepPartial<S>> : T extends object ? DeepPartial<T> : never): R;
toBe<E = T>(expected: E): R;
toEqual<E = T>(expected: E): R;
```

This helps autocompleting loads when writing tests.

[a comment by an automated account is not shown]

**Comment by tonyhallett** [comment]
Plese see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47020 that was reverted.

**Reply by tomhicks** [reply]
meh, there's pretty much one person moaning that they don't want to change their tests, despite very good reasons given (production breakage: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47020#issuecomment-738281883) to have strict typing in your tests.

There are some wrinkles to iron out, sure, but that's throwing the baby out with the bathwater.

**Reply by tonyhallett** [reply]
> meh, there's pretty much one person moaning that they don't want to change their tests,

There were others and their solid reasoning has been provided.

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47020#issuecomment-740569315
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47020#issuecomment-738288238
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47020#issuecomment-739139375


https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47553#issuecomment-692270723
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47553#issuecomment-692248028

I also think is perfectly acceptable for a mock to return a fake type and to later expect a value of the real type toBe equal to it.

> despite very good reasons given (production breakage: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47020#issuecomment-738281883) to have strict typing in your tests.

There is no denying that changing the typing would have benefits but the counter arguments are strong.  
The pr has not been reverted due to a reluctance to change in spite of very good reasons for the change.

> There are some wrinkles to iron out, sure, but that's throwing the baby out with the bathwater.

The wrinkles that are the arguments against enforcing strict typing will remain.

Perhaps a future pr will be approved and all those in disagreement from the millions of Jest Definitely Typed users will have to provide the generic type where necessary.

Also see [comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/50867#issuecomment-838874240)
